crates/natrix_macros/src/lib.rs

```diff
@@ -24,6 +24,369 @@ pub fn format_elements(input: proc_macro::TokenStream) -> proc_macro::TokenStrea
+                    });
+                } else {
+                    let none_field_inits: Vec<_> = field_names.iter().map(|name| quote!(#name: ::natrix::access::Ref::FaillableMut(None))).collect();
+                    faillable_none_arms.push(quote! {
+                        #projected_name::#variant_name { #(#none_field_inits),* }
+                    });
+                }
+
+                faillable_some_arms.push(quote! {
+                    #name::#variant_name { #(#field_names),* } => #projected_name::#variant_name { #(#field_names: ::natrix::access::Ref::FaillableMut(Some(#field_names))),* }
+                });
+            },
+            syn::Fields::Unnamed(fields) => {
+                let field_types: Vec<_> = fields.unnamed.iter().map(|f| &f.ty).collect();
+                let field_names: Vec<_> = (0..fields.unnamed.len()).map(|i| format_ident!("field_{}", i)).collect();
+                
+                projected_variants.push(quote! {
+                    #variant_name(#(::natrix::access::Ref<'a, #field_types>),*)
+                });
+
+                read_arms.push(quote! {
+                    #name::#variant_name(#(#field_names),*) => #projected_name::#variant_name(#(::natrix::access::Ref::Read(#field_names)),*)
+                });
+
+                mut_arms.push(quote! {
+                    #name::#variant_name(#(#field_names),*) => #projected_name::#variant_name(#(::natrix::access::Ref::Mut(#field_names)),*)
+                });
+
+                if fields.unnamed.is_empty() {
+                    faillable_none_arms.push(quote! {
+                        #projected_name::#variant_name
+                    });
+                } else {
+                    let none_refs: Vec<_> = (0..fields.unnamed.len()).map(|_| quote!(::natrix::access::Ref::FaillableMut(None))).collect();
+                    faillable_none_arms.push(quote! {
+                        #projected_name::#variant_name(#(#none_refs),*)
+                    });
+                }
+
+                faillable_some_arms.push(quote! {
+                    #name::#variant_name(#(#field_names),*) => #projected_name::#variant_name(#(::natrix::access::Ref::FaillableMut(Some(#field_names))),*)
+                });
+            },
+            syn::Fields::Unit => {
+                projected_variants.push(quote! {
+                    #variant_name
+                });
+
+                read_arms.push(quote! {
+                    #name::#variant_name => #projected_name::#variant_name
+                });
+
+                mut_arms.push(quote! {
+                    #name::#variant_name => #projected_name::#variant_name
+                });
+
+                faillable_none_arms.push(quote! {
+                    #projected_name::#variant_name
+                });
+
+                faillable_some_arms.push(quote! {
+                    #name::#variant_name => #projected_name::#variant_name
+                });
+            }
```

Comment:
We have the `get_fields` helper for a reason here.
you could prolly modify that a bit to work here.